feat: Add confirmation email for form respondents

dtretyakov opened the pull request:
Implements issue #525 - Send confirmation emails to form respondents after submission.

Features:
- Add database migration for confirmation email settings (enabled, subject, body)
- Add confirmation email fields to Form entity
- Implement email sending with placeholder replacement:
  - {formTitle}, {formDescription} placeholders
  - Field name placeholders (e.g. {name}, {email})
  - Automatic data overview if {data} not in template
- Add UI in Settings sidebar to configure confirmation emails
- Automatically detect email field from form submissions
- Send email using system mail account

Technical changes:
- Add sendConfirmationEmail() method to FormsService
- Integrate email sending into notifyNewSubmission() flow
- Add unit tests for confirmation email functionality
- Update FormsService constructor with IMailer and AnswerMapper dependencies
- Update documentation (DataStructure.md, CHANGELOG.en.md)

The feature requires at least one email-validated short text question in the form. Email sending failures are logged but don't break the submission process.

<img width="837" height="862" alt="Screenshot 2025-12-19 105758" src="https://github.com/user-attachments/assets/fad62e98-fd35-4cc6-99af-4c454ff33466" />

Chartman123 commented:
@dtretyakov I'll have a closer look in the next days :)

susnux reviewed (commented):
Looks quite clean but did not review properly yet.
Will do soon!

Chartman123 commented:
@dtretyakov Could you please move your changes to a branch other than `main` in your fork? This makes it easier to check out your changes in my local dev environment

susnux reviewed (commented):
Overall this makes sense to me, but from experience with other tables like this it turned out that its not ideal for scaling.

The problem is that many things in the table are only used by few tables but cause huge size of the table like "submission message" or from this PR "confirmation mail body".

Its not a problem introduced with this PR - so we can also do this later!
But just for some feedback here cc @Chartman123 

I wonder if we should first - or later - split the table to get some "forms metadata" or similar so that we can put data not needed for every action there.
That would reduce database select latency a lot for large instances (e.g. this information is only needed on *edit* or *insert submit* but not for the majority of requests which would just load the form for submitting.

susnux reviewed (commented):
So for the pull request itself I think it is quite nice! Good work :)

Just about the email field I am not sure if we should not better explicitly let the creator flag the input field to use.

Chartman123 commented:
> let the creator flag the input field to use.

Could be part of the `extraSettings`

dtretyakov committed:
feat: Add confirmation email for form respondents

Implements issue #525 - Send confirmation emails to form respondents after submission.

Features:
- Add confirmation email settings (enabled, subject, body) to Form entity.
- Implement email sending with placeholder replacement:
  - {formTitle}, {formDescription}
  - Field placeholders based on name or label (e.g. {name}, {email}).
- Allow form creators to explicitly select the recipient email field when multiple email fields are present.
- Add UI in Settings sidebar to configure confirmation emails and recipient selection.
- Replace basic email validation with Nextcloud's internal IEmailValidator.
- Integration with activity notifications and background jobs for file syncing.

Technical changes:
- Database migration for new Form properties.
- Enhanced FormsService with email sending logic and validation.
- Extensive unit and integration tests covering the new functionality.
- Updated API documentation and OpenAPI spec.

Signed-off-by: Dmitry Tretyakov <dtretyakov@gmail.com>

dtretyakov commented on lib/Service/FormsService.php:
```diff
@@ -735,6 +739,151 @@ public function notifyNewSubmission(Form $form, Submission $submission): void {
+			}
+
+			$emailValue = $answerMap[$questionId][0];
+			if ($this->mailer->validateMailAddress($emailValue)) {
```
switched to `IEmailValidator`

dtretyakov commented on lib/Service/FormsService.php:
```diff
@@ -735,6 +739,151 @@ public function notifyNewSubmission(Form $form, Submission $submission): void {
+				'submissionId' => $submission->getId(),
+				'emailCount' => count($recipientEmails),
+			]);
+		}
```
Added a dedicated flag, stored in extra settings

dtretyakov commented:
> Could you please move your changes to a branch other than main in your fork? 

@Chartman123 I created a draft PR with the same changes https://github.com/nextcloud/forms/pull/3289

> Could be part of the extraSettings

Added the dedicated flag `confirmationEmailRecipient` stored in `extraSettings`

Chartman123 commented:
@dtretyakov ok, so we can close this one?

dtretyakov committed:
refactor: streamline confirmation email recipient setup

dtretyakov commented:
Closed in favor of https://github.com/nextcloud/forms/pull/3289
